clash-protocols/src/Protocols/PacketStream/Base.hs

```diff
@@ -0,0 +1,554 @@
+mapMetaS fS = Circuit $ \(fwdIn, bwdIn) -> (bwdIn, go <$> bundle (fwdIn, fS))
+ where
+  go (inp, f) = (\inPkt -> inPkt{_meta = f (_meta inPkt)}) <$> inp
+
+-- | Like 'Data.List.filter', but over the metadata of a 'PacketStream'.
+filterMeta ::
+  -- | Predicate which specifies whether to keep a fragment based on its metadata
+  (meta -> Bool) ->
+  Circuit (PacketStream dom dataWidth meta) (PacketStream dom dataWidth meta)
+filterMeta p = filterMetaS (pure p)
+
+-- | Like 'filterMeta', but can reason over signals.
+filterMetaS ::
+  -- | Predicate which specifies whether to keep a fragment based on its metadata,
+  --   wrapped in a @Signal@
+  Signal dom (meta -> Bool) ->
+  Circuit (PacketStream dom dataWidth meta) (PacketStream dom dataWidth meta)
+filterMetaS pS = Circuit $ \(fwdIn, bwdIn) -> unbundle (go <$> bundle (fwdIn, bwdIn, pS))
+ where
+  go (Nothing, bwdIn, _) = (bwdIn, Nothing)
+  go (Just inPkt, bwdIn, predicate)
+    | predicate (_meta inPkt) = (bwdIn, Just inPkt)
+    | otherwise = (PacketStreamS2M True, Nothing)
+
+-- | Like 'Data.Either.either', but over the metadata of a 'PacketStream'.
+eitherMeta ::
+  (a -> c) ->
+  (b -> c) ->
+  Circuit
+    (PacketStream dom dataWidth (Either a b))
+    (PacketStream dom dataWidth c)
+eitherMeta f g = eitherMetaS (pure f) (pure g)
+
+-- | Like 'eitherMeta', but can reason over signals.
+eitherMetaS ::
+  Signal dom (a -> c) ->
+  Signal dom (b -> c) ->
+  Circuit
+    (PacketStream dom dataWidth (Either a b))
+    (PacketStream dom dataWidth c)
+eitherMetaS fS gS = mapMetaS (liftA2 P.either fS gS)
+
+-- | Like 'Data.Bifunctor.bimap', but over the metadata of a 'PacketStream'.
+bimapMeta ::
+  (B.Bifunctor p) =>
+  (a -> b) ->
+  (c -> d) ->
+  Circuit
+    (PacketStream dom dataWidth (p a c))
+    (PacketStream dom dataWidth (p b d))
+bimapMeta f g = bimapMetaS (pure f) (pure g)
+
+-- | Like 'bimapMeta', but can reason over signals.
+bimapMetaS ::
+  (B.Bifunctor p) =>
+  Signal dom (a -> b) ->
+  Signal dom (c -> d) ->
+  Circuit
+    (PacketStream dom dataWidth (p a c))
+    (PacketStream dom dataWidth (p b d))
+bimapMetaS fS gS = mapMetaS (liftA2 B.bimap fS gS)
+
+-- | Like 'Data.Bifunctor.first', but over the metadata of a 'PacketStream'.
+firstMeta ::
+  (B.Bifunctor p) =>
+  (a -> b) ->
+  Circuit
+    (PacketStream dom dataWidth (p a c))
+    (PacketStream dom dataWidth (p b c))
+firstMeta f = firstMetaS (pure f)
+
+-- | Like 'firstMeta', but can reason over signals.
+firstMetaS ::
+  (B.Bifunctor p) =>
+  Signal dom (a -> b) ->
+  Circuit
+    (PacketStream dom dataWidth (p a c))
+    (PacketStream dom dataWidth (p b c))
+firstMetaS fS = mapMetaS (B.first <$> fS)
+
+-- | Like 'Data.Bifunctor.second', but over the metadata of a 'PacketStream'.
+secondMeta ::
+  (B.Bifunctor p) =>
+  (b -> c) ->
+  Circuit
+    (PacketStream dom dataWidth (p a b))
+    (PacketStream dom dataWidth (p a c))
+secondMeta f = secondMetaS (pure f)
+
+-- | Like 'secondMeta', but can reason over signals.
```

Comment:
Perhaps as @DigitalBrains1 noted we should make clear that these `*S` circuit combinators are akin to `<*>`.